tests/test_bootstrap.py

```diff
@@ -0,0 +1,556 @@
+        successfully calls once the method `endure_dir`
+        """
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+
+        bs.prepare_dist_dir("fake_name")
+        mock_ensure_dir.assert_called_once_with(bs.dist_dir)
+
+    @mock.patch("pythonforandroid.bootstrap.open", create=True)
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy")
+    @mock.patch("pythonforandroid.bootstrap.os.makedirs")
+    def test_bootstrap_prepare_build_dir(
+        self, mock_os_makedirs, mock_shutil_copy, mock_chdir, mock_open
+    ):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_build_dir`
+        successfully calls the methods that we need to prepare a build dir.
+        """
+
+        # prepare bootstrap
+        bs = Bootstrap().get_bootstrap("service_only", self.ctx)
+        self.ctx.bootstrap = bs
+
+        # test that prepare_build_dir runs (notice that we mock
+        # any file/dir creation so we can speed up the tests)
+        bs.prepare_build_dir()
+
+        # make sure that the open command has been called only once
+        mock_open.assert_called_once_with("project.properties", "w")
+        mock_open.reset_mock()
+
+        # check that the other mocks we made are actually called
+        mock_os_makedirs.assert_called()
+        mock_shutil_copy.assert_called()
+        mock_chdir.assert_called()
+
+    @mock.patch("pythonforandroid.bootstrap.os.path.isfile")
+    @mock.patch("pythonforandroid.bootstrap.os.path.exists")
+    @mock.patch("pythonforandroid.bootstrap.os.unlink")
+    @mock.patch("pythonforandroid.bootstrap.open", create=True)
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.ln")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.mkdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.isdir")
+    def test_bootstrap_prepare_build_dir_with_java_src(
+        self,
+        mock_isdir,
+        mock_sh_rm,
+        mock_sh_mkdir,
+        mock_listdir,
+        mock_sh_ln,
+        mock_chdir,
+        mock_open,
+        mock_os_unlink,
+        mock_os_path_exists,
+        mock_os_path_isfile,
+    ):
+        """A test which will initialize a bootstrap and will check perform
+        another test for method
+        :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_build_dir`. In
+        here we will simulate that we have `with_java_src` set to some value.
+        """
+        self.ctx.symlink_java_src = ["some_java_src"]
+        mock_isdir.return_value = False
+        mock_listdir.return_value = [
+            "jnius",
+            "kivy",
+            "Kivy-1.11.0.dev0-py3.7.egg-info",
+            "pyjnius-1.2.1.dev0-py3.7.egg",
+        ]
+        mock_sh_ln.return_value.return_value = mock.Mock()
```

Comment:
Why is this for? Basically it would make the `ln()()` call return a `Mock`, but it almost does that by default since it's a `MagicMock` hence `ln()()` returns a `MagicMock` even if you remove that line. So I'm not sure why this is needed :thinking: